README.md
# PyWeekends: Weekend Dates Formatter

PyWeekends is a Python package that provides a convenient way to generate weekend dates based on the Calendar. 
It includes a class named `Weekend` to generate weekend dates and an `OutputFormatter` class to format 
the generated dates into JSON.

## Features

- Generate weekend dates for all months in a given year.
- Generate weekend dates for a specific month.
- Format the generated dates as JSON.

## Installation

To install the package, use pip:

```bash
pip install PyWeekends
```
## Usage
```python
from PyWeekends import Weekend, WeekendStart

# Generate weekend dates for a specific month and format as JSON
w = Weekend(year=2024)
january_weekends = w.month_weekends("January")
json_data = january_weekends.as_json()
print(json_data)
# Output
# ["6", "7", "13", "14", "20", "21", "27", "28"]
# Generate weekend dates for all months and format as JSON
all_month_weekends = w.all_weekends()
json_data_all_months = all_month_weekends.as_json()
print(json_data_all_months)

#Output
# {"1": ["6", "7", "13", "14", "20", "21", "27", "28"], "2": ["3", "4", "10", "11", "17", "18", "24", "25"], 
# "3": ["2", "3", "9", "10", "16", "17", "23", "24", "30", "31"], "4": ["6", "7", "13", "14", "20", "21", "27", "28"], 
# "5": ["4", "5", "11", "12", "18", "19", "25", "26"], "6": ["1", "2", "8", "9", "15", "16", "22", "23", "29", "30"],
# "7": ["6", "7", "13", "14", "20", "21", "27", "28"], "8": ["3", "4", "10", "11", "17", "18", "24", "25", "31"], 
# "9": ["1", "7", "8", "14", "15", "21", "22", "28", "29"], "10": ["5", "6", "12", "13", "19", "20", "26", "27"], 
# "11": ["2", "3", "9", "10", "16", "17", "23", "24", "30"], "12": ["1", "7", "8", "14", "15", "21", "22", "28", "29"]}

```

## Changelog
See [Changelog](CHANGELOG.md)

## License
[MIT](LICENSE)

CHANGELOG.md
## Changelog

All notable changes to this project will be documented in this file. <br/>
The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/). This project adheres
to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

tests/test_weekend.py
import unittest
import json
from PyWeekends import Weekend, WeekendStart


class TestWeekend(unittest.TestCase):

    def test_all_weekends(self):
        w = Weekend(year=2024)
        weekends = w.all_weekends(enable_month_name=True).as_json()
        expected_weekends = {
            'January': ['6', '7', '13', '14', '20', '21', '27', '28'],
            'February': ['3', '4', '10', '11', '17', '18', '24', '25'],
            'March': ['2', '3', '9', '10', '16', '17', '23', '24', '30', '31'],
            'April': ['6', '7', '13', '14', '20', '21', '27', '28'],
            'May': ['4', '5', '11', '12', '18', '19', '25', '26'],
            'June': ['1', '2', '8', '9', '15', '16', '22', '23', '29', '30'],
            'July': ['6', '7', '13', '14', '20', '21', '27', '28'],
            'August': ['3', '4', '10', '11', '17', '18', '24', '25', '31'],
            'September': ['1', '7', '8', '14', '15', '21', '22', '28', '29'],
            'October': ['5', '6', '12', '13', '19', '20', '26', '27'],
            'November': ['2', '3', '9', '10', '16', '17', '23', '24', '30'],
            'December': ['1', '7', '8', '14', '15', '21', '22', '28', '29']
        }

        self.assertEqual(json.loads(weekends), expected_weekends)

    def test_month_weekends(self):
        w = Weekend(year=2024, weekend_start=WeekendStart.FRIDAY)
        january_weekends = w.month_weekends('January').as_json()
        expected_january_weekends = ['5', '6', '12', '13', '19', '20', '26', '27']
        self.assertEqual(json.loads(january_weekends), expected_january_weekends)

    def test_month_weekends_invalid_month(self):
        w = Weekend(year=2024, weekend_start=WeekendStart.FRIDAY)
        with self.assertRaises(ValueError):
            w.month_weekends('InvalidMonth')

    def test_month_weekends_invalid_month_case(self):
        w = Weekend(year=2024, weekend_start=WeekendStart.FRIDAY)
        with self.assertRaises(ValueError):
            w.month_weekends('jaNuARy4')  # Invalid case

    def test_month_weekends_invalid_year(self):
        # Test with a year in the future
        w = Weekend(year=454545, weekend_start=WeekendStart.FRIDAY)
        with self.assertRaises(ValueError):
            w.month_weekends('January')


tests/test_output_formatter.py
import json
import unittest

from PyWeekends.output_formatter import OutputFormatter


class TestOutputFormatter(unittest.TestCase):

    def test_as_json(self):
        data = {'January': ['5', '6', '12', '13', '19', '20', '26', '27']}
        formatter = OutputFormatter(data)
        expected_json = json.dumps(data)
        self.assertEqual(formatter.as_json(), expected_json)

    def test_as_json_invalid(self):
        # Test with non-dictionary or non-list data
        formatter = OutputFormatter('invalid_data')
        with self.assertRaises(TypeError):
            formatter.as_json()

        # Test with invalid JSON data
        formatter = OutputFormatter({"January", "February"})
        with self.assertRaises(TypeError):
            formatter.as_json()


PyWeekends/weekend.py
import calendar
import datetime
from typing import List, AnyStr
from enum import Enum

from PyWeekends.output_formatter import OutputFormatter


class WeekendStart(Enum):
    FRIDAY = 4
    SATURDAY = 5
    SUNDAY = 6


class Weekend:
    """
    The Weekend class generates weekend dates based on the Calendar.

    Attributes:
        year (int): The year for which to generate weekend dates. Defaults to the current year.
        weekend_start (WeekendStart): The day at which the weekend starts (Friday, Saturday, or Sunday).
            Defaults to Saturday.

    Methods:
        all_weekends(enable_month_name: bool = False) -> OutputFormatter:
            Generates weekend dates for all months in the given year.
            Args:
                enable_month_name (bool): True to return month names, False to return month numbers.
            Returns:
                OutputFormatter: An instance of OutputFormatter containing the weekend dates.

        Month_weekends(month_name: AnyStr) -> List[str]:
            Generates weekend dates for a specific month.
            Args:
                month_name (str): Name of the month (case-insensitive).
            Returns:
                OutputFormatter: An instance of OutputFormatter containing the weekend dates.

        __Weekend_dates(month: int) -> List[str]:
            Helper method to generate weekend dates for a specific month.
            Args:
                month (int): Month number.
            Returns:
                List[str]: List of weekend dates for the specified month.

    Examples:
        w = Weekend(year=2024, weekend_start=WeekendStart.FRIDAY)
        weekends = w.all_weekends().as_json()
        print(weekends) # Outputs the weekend dates for all months in JSON format
    """

    def __init__(self, year: int = None, weekend_start: WeekendStart = WeekendStart.SATURDAY):
        """
        Initializes a Weekend instance.
        Args:
            year (int): The year for which to generate weekend dates. Defaults to the current year.
            weekend_start (WeekendStart): The day at which the weekend starts (Friday, Saturday, or Sunday).
            Defaults to Saturday.
        """
        self.year = year or datetime.datetime.now().year
        self.weekend_start = weekend_start

    def all_weekends(self, enable_month_name: bool = False) -> OutputFormatter:
        """
        Returns weekend dates for all months.
        :param enable_month_name: bool: True for month names, False for month numbers.
        :return: {"1": [<weekend date>, <weekend date>] ...}
        :rtype: dict
        """
        return OutputFormatter(
            {str(calendar.month_name[month] if enable_month_name else month): self.__weekend_dates(month) for month
             in range(1, 13)})

    def month_weekends(self, month_name: AnyStr) -> OutputFormatter:
        """
        Returns weekend dates for a specific month.
        :param month_name: str: Name of the month (case-insensitive).
        :return: List of weekend dates.
        """
        month = next((k for k, v in enumerate(calendar.month_name) if v.lower() == month_name.lower()), None)
        if not month:
            raise ValueError(f"Invalid month name: {month_name}")
        return OutputFormatter(self.__weekend_dates(month))

    def __weekend_dates(self, month: int) -> List[str]:
        """
        Returns weekend dates for a specific month.
        :param month: int: Month number.
        :return: List of weekend dates.
        """
        date_range = range(1, calendar.monthrange(year=self.year, month=month)[1] + 1)
        if self.weekend_start == WeekendStart.FRIDAY:
            days = [str(day) for day in date_range if datetime.date(self.year, month, day).weekday() in (4, 5)]
        else:
            days = [str(day) for day in date_range if datetime.date(self.year, month, day).weekday() in (5, 6)]

        return days


PyWeekends/output_formatter.py
import json
from typing import List, Dict, Union


class OutputFormatter:
    """
    The OutputFormatter class formats data and provides methods to represent it as JSON or other formats.

    Attributes:
        data (Union[Dict[str, List[str]], List]): The data to be formatted.

    Methods:
        as_json() -> str:
            Formats the data as JSON and returns the JSON string.

        to_original -> Union[Dict[str, List[str]], List]:
            Formats the data as an original object.e.g dict or List

    Example usage:
        data = {'January': ['5', '6', '12', '13', '19', '20', '26', '27'],
        'February': ['2', '3', '9', '10', '16', '17', '23', '24']}
        formatter = OutputFormatter(data)
        json_data = formatter.as_json()
        print(json_data)  # Outputs the data in JSON format

    """

    def __init__(self, data: Union[Dict[str, List[str]], List]):
        """
        Initializes an OutputFormatter instance.

        Args:
            data (Union[Dict[str, List[str]], List]): The data to be formatted.
        """
        self.data = data

    def as_json(self) -> str:
        """
        Formats the data as JSON and returns the JSON string.

        Returns:
            str: The data in JSON format.
        """
        if not isinstance(self.data, (dict, list)):
            raise TypeError("Invalid data type. Expected dict or list.")
        return json.dumps(self.data)

    def to_original(self) -> Union[Dict[str, List[str]], List]:
        """
        Formats the data as Original of parent class.
        Returns:
            Union[Dict[str, List[str]], List]
        """
        return self.data
